### PR TITLE
Add location to linkage symbols

### PR DESCRIPTION
**Pull request by AsterMiha**
Currently:
CPPMangleDeclaration - location of `struct` or `class` - changed to location of  `extern`
CPPNamespaceDeclaration - location of `extern` (same as `Nspace`)
LinkDeclaration - location of `extern`

### CODE REVIEW

**Review by RazvanN7 — APPROVED** [review]

File: test/unit/parser/linkage_location.d [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+    {
+        l = ld.loc;
+    }
+}
+
+enum Links = [
+    "C",
+    "C++",
+    "D",
+    "Objective-C"];
```
Comment (jacob-carlborg) [inline test/unit/parser/linkage_location.d:219]: Should include `Windows` and `System` for completeness.

**Comment by jacob-carlborg** [comment]
> CPPMangleDeclaration - location of struct or class

All locations should start at the first byte of the AST node. It's also inconsistent compared to `CPPNamespaceDeclaration` and `LinkDeclaration`.

**Comment by RazvanN7** [comment]
@jacob-carlborg is the current state of this PR enough for your approval?

**Comment by jacob-carlborg** [comment]
> is the current state of this PR enough for your approval?

Yes, I didn't know there had been changes.

**Review by jacob-carlborg — APPROVED** [review]

File: src/dmd/parse.d [hunk 2]
```diff
@@ -4253,7 +4255,8 @@ final class Parser(AST) : Lexer
     }
 
     private void parseStorageClasses(ref StorageClass storage_class, ref LINK link,
-        ref bool setAlignment, ref AST.Expression ealign, ref AST.Expressions* udas)
+        ref bool setAlignment, ref AST.Expression ealign, ref AST.Expressions* udas,
+        ref Loc linkloc)
```
Comment (jacob-carlborg) [inline src/dmd/parse.d:4259]: Should this be `out`, or is it used for input as well?

**Comment by jacob-carlborg** [comment]
@AsterMiha when you made a change to address a comment (or a batch of comments), please write a new comment to indicate that you have addressed the comment. It's much easier to notice.

File: src/dmd/parse.d (hunk 2 above)
Comment (AsterMiha) [reply src/dmd/parse.d]: You're right. Changed it from `ref` to `out`

**Comment by Geod24** [comment]
@AsterMiha : What's the status on this ?

**Comment by AsterMiha** [comment]
@Geod24 Should be done. I'll try a rebase to see if it fixes the failed tests.

**Comment by RazvanN7** [comment]
@AsterMiha please rebase this and let's get it in.

File: src/dmd/frontend.h [hunk 3]
```diff
@@ -2024,8 +2024,8 @@ class LinkDeclaration final : public AttribDeclaration
 {
 public:
     LINK linkage;
-    static LinkDeclaration* create(LINK p, Array<Dsymbol* >* decl);
-    LinkDeclaration* syntaxCopy(Dsymbol* s);
+    static LinkDeclaration* create(const Loc& loc, LINK p, Array<Dsymbol* >* decl);
+    Dsymbol* syntaxCopy(Dsymbol* s);
```
Comment (RazvanN7) [inline src/dmd/frontend.h:2028]: This is the reason why the tests are failing. It seems that the return of `syntaxCopy` was not modified in the D file. Why is it modified in the header files?

File: src/dmd/attrib.d [hunk 4]
```diff
@@ -393,22 +393,22 @@ extern (C++) final class LinkDeclaration : AttribDeclaration
...
     }
 
-    static LinkDeclaration create(LINK p, Dsymbols* decl)
+    static LinkDeclaration create(const ref Loc loc, LINK p, Dsymbols* decl)
     {
-        return new LinkDeclaration(p, decl);
+        return new LinkDeclaration(loc, p, decl);
     }
 
     override LinkDeclaration syntaxCopy(Dsymbol s)
```
Comment (RazvanN7) [inline src/dmd/attrib.d:408]: This function returns a LinkDeclaration, why is it returning a Dsymbol in the header files?

**Comment by MoonlightSentinel** [comment]
@AsterMiha Offtopic question, but do you intent to add a location to `Parameter` as well?

**Pull request merged**